### PR TITLE
NEW Read only support for the React components in this module

### LINKED ISSUE #159: Archiving a page causes an error
Slightly misleading title, please forgive.

Steps to recreate:

4.3.x-dev project, `silverstripe/ckan-registry` installed, on [current `dev-master`](https://github.com/silverstripe/silverstripe-ckan-registry/tree/bc06361b8e98ac1098c91ba7b5c1525b6d3161ee) (bc06361b8)
- Log in to CMS
- Create CKAN Registry Page
- Fill it with data (e.g. [Auckland family services](https://catalogue.data.govt.nz/dataset/family-services-directory/resource/496d2b91-2e9a-4e72-9c07-cf41141aefa6))
- Save and publish
- Reload page
- Select "more options" three dots next to publish button
- Click "Unpublish and archive"A
- Wait for loading to complete
- Refresh page (F5/Ctrl+R/Cmd+r)
- Observe `nl2br` error about expecting a string, getting an array.

Tested with `composer create-project silverstripe/installer cmsarchiveandrefreshbug 4.3.x-dev` and could not reproduce (attempting to `Unpublish & archive` the "Contact us" page). This indicates one of two things to me:
1. The issue is with this module
2. The issue is with any page that has a react rendered form field on it

Unsure which, maybe neither. Logged issue here.

### LINKED ISSUE #179: Deselecting "Select from all options" makes checkboxes disappear
I can't exactly remember what we decided when deselecting the checkboxes (if they stay where they are or move to the top of the list). We could leave them where they are and have a "re/deselect all" see #100. For now I think they should move to the top 🤔 cc @ScopeyNZ 

deselect

### PR DESCRIPTION
**Pull request by ScopeyNZ**
- Adds support for the readOnly flag in the React components for "PresentedOptions", "ResourceLocator", and "ResultConditions"
- Updates those fields to only update the read only property when performing a read only transformation
- CKANResourceLocatorField now uses a TextField from core rather than using a pointless URIInput component
- Removed in-line use of fieldHolder HOC in ResourceLocatorField
- Fixed a bug where checkbox items disappeared when selecting them in the PresentedOptionsField while it was new
- Updates the `GridFieldResourceTitle` GridField component to be a "read only" component but not show the additional button while read only.

Fixes #179
Fixes #159

### CODE REVIEW

File: client/src/components/CKANPresentedOptionsField.scss [hunk 1]
```diff
@@ -50,4 +50,11 @@
   .form__field-description {
     font-style: italic;
   }
+
+  // Override read-only styling
+  .readonly .form__field-holder & {
+    background-color: #f7f8fa;
```
Comment (robbieaverill) [inline client/src/components/CKANPresentedOptionsField.scss]: Any chance of a variable here? You've used $body-bg further down in a similar block

File: client/src/components/CKANResourceLocatorField.js [hunk 2]
```diff
@@ -351,37 +363,46 @@ class CKANResourceLocatorField extends Component {
...
     const invalidMessage = this.getInvalidURLMessage();
     // And determine validity by the existance of that message
     const invalid = !!invalidMessage;
 
-    // Props for the URLInput
     const inputProps = {
+      name: `${name}-uri`,
       title: i18n._t('CKANResourceLocatorField.DATA_SOURCE_URL', 'Data source URL'),
-      extraClass: 'form-field--no-divider stacked',
+      // extraClass: 'form-field--no-divider stacked',
```
Comment (robbieaverill) [inline client/src/components/CKANResourceLocatorField.js]: Remove?

File: src/Forms/PresentedOptionsField.php [hunk 3]
```diff
@@ -82,6 +89,11 @@ public static function getSelectTypes()
         ];
     }
 
+    public function performReadonlyTransformation()
+    {
+        return $this->setReadonly(true);
```
Comment (robbieaverill) [inline src/Forms/PresentedOptionsField.php]: Usually these methods are immutable - might be worth returning a clone of $this as well

File: src/Forms/PresentedOptionsField.php [hunk 4]
```diff
@@ -99,4 +111,20 @@ public function setResource(Resource $resource)
...
+    {
+        return $this->selectedFields;
+    }
+
+    /**
+     * @param array $selectedFields
+     */
+    public function setSelectedFields($selectedFields)
+    {
+        $this->selectedFields = $selectedFields;
```
Comment (robbieaverill) [inline src/Forms/PresentedOptionsField.php]: Add fluent return

File: src/Forms/PresentedOptionsField.php [hunk 5]
```diff
@@ -99,4 +111,20 @@ public function setResource(Resource $resource)
...
+     */
+    public function getSelectedFields()
+    {
+        return $this->selectedFields;
+    }
+
+    /**
+     * @param array $selectedFields
+     */
+    public function setSelectedFields($selectedFields)
```
Comment (robbieaverill) [inline src/Forms/PresentedOptionsField.php]: Type hint array?

File: src/Forms/ResourceLocatorField.php [hunk 6]
```diff
@@ -80,6 +80,15 @@ public function getSchemaDataDefaults()
         ]);
     }
 
+    public function performReadonlyTransformation()
+    {
+        // Clear out the description that's only relevant when the field is editable.
+        $this->setDescription('');
+
+        return $this->setReadonly(true);
```
Comment (robbieaverill) [inline src/Forms/ResourceLocatorField.php]: Similar to previous suggestion: make this immutable

File: src/Forms/ResultConditionsField.php [hunk 7]
```diff
@@ -40,6 +40,11 @@ public function Type()
         return 'ckan-result-conditions';
     }
 
+    public function performReadonlyTransformation()
+    {
+        return $this->setReadonly(true);
```
Comment (robbieaverill) [inline src/Forms/ResultConditionsField.php]: And here

File: client/src/components/CKANPresentedOptionsField.scss (hunk 1 above)
Comment (ScopeyNZ) [reply client/src/components/CKANPresentedOptionsField.scss]: 🤦‍♂️

File: src/Forms/PresentedOptionsField.php (hunk 5 above)
Comment (ScopeyNZ) [reply src/Forms/PresentedOptionsField.php]: In the end this wasn't required. I'll just remove it.

File: src/Forms/ResourceLocatorField.php [hunk 8]
```diff
@@ -80,6 +80,15 @@ public function getSchemaDataDefaults()
         ]);
     }
 
+    public function performReadonlyTransformation()
+    {
+        // Clear out the description that's only relevant when the field is editable.
+        $this->setDescription('');
```
Comment (robbieaverill) [inline src/Forms/ResourceLocatorField.php]: This should be done on the cloned instance rather than the current

**Comment by robbieaverill** [comment]
We're removing some API so switched to major

**Comment by robbieaverill** [comment]
I don't think the display conditions field is working. It's either not showing up at all for me or showing the label with an empty field content.

Console errors:

```
Uncaught exception TypeError: Cannot read property 'title' of undefined
    at t.value (bundle-admin.js?m=1551138705:1)
    at t.value (bundle-admin.js?m=1551138705:1)
    at c._renderValidatedComponentWithoutOwnerOrContext (vendor.js?m=1549036726:1)
    at c._renderValidatedComponent (vendor.js?m=1549036726:1)
    at c.performInitialMount (vendor.js?m=1549036726:1)
    at c.mountComponent (vendor.js?m=1549036726:1)
    at Object.mountComponent (vendor.js?m=1549036726:1)
    at c.performInitialMount (vendor.js?m=1549036726:1)
    at c.mountComponent (vendor.js?m=1549036726:1)
    at Object.mountComponent (vendor.js?m=1549036726:1) in onmatch on <div type=​"text" name=​"DisplayConditions" value=​"{"0":​{"match-select":​"0","match-text":​"sdfsdf"}​}​" class=​"ckan-result-conditions ckan-result-conditions__container" id=​"Form_ItemEditForm_DisplayConditions" readonly=​"readonly" data-schema=​"{"name":​"DisplayConditions","id":​"Form_ItemEditForm_DisplayConditions","type":​"text","schemaType":​"Text","component":​"ResultConditions","holderId":​"Form_ItemEditForm_DisplayConditions_Holder","title":​"Result conditions","source":​null,"extraClass":​"ckan-result-conditions ckan-result-conditions__container","description":​null,"rightTitle":​null,"leftTitle":​null,"readOnly":​true,"disabled":​false,"customValidationMessage":​"","validation":​[]​,"attributes":​[]​,"autoFocus":​false,"data":​{"maxlength":​null,"source":​[{"value":​1,"title":​"Must match"}​,{"value":​0,"title":​"Must not match"}​]​,"matchTypeDefault":​1}​}​">​…​</div>​

...
vendor.js?m=1549036726:1 Stack Trace:
TypeError: Cannot read property 'title' of undefined
    at t.value (http://ss43.localhost/resources/vendor/silverstripe/ckan-registry/client/dist/js/bundle-admin.js?m=1551138705:1:22603)
    at t.value (http://ss43.localhost/resources/vendor/silverstripe/ckan-registry/client/dist/js/bundle-admin.js?m=1551138705:1:22773)
    at c._renderValidatedComponentWithoutOwnerOrContext (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1438958)
    at c._renderValidatedComponent (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1439085)
    at c.performInitialMount (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1434916)
    at c.mountComponent (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1433966)
    at Object.mountComponent (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1497578)
    at c.performInitialMount (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1435077)
    at c.mountComponent (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1433966)
    at Object.mountComponent (http://ss43.localhost/resources/vendor/silverstripe/admin/client/dist/js/vendor.js?m=1549036726:1:1497578)
```

File: client/src/components/CKANResultConditionsField.js [hunk 9]
```diff
@@ -155,7 +155,28 @@ class CKANResultConditionsField extends Component {
...
+   *
+   * @returns {HTMLElement|null}
+   */
+  renderReadOnly() {
+    const { data: { source } } = this.props;
+    const input = this.getInputValue();
+    const type = source.find(candidate => candidate.value === this.getSelectValue()).title;
+
+    if (!type || !input) {
+      return null;
```
Comment (robbieaverill) [inline client/src/components/CKANResultConditionsField.js]: This is a little weird in the UI in my opinion. Would be better to render an empty UI control maybe

Comment (robbieaverill) [reply client/src/components/CKANResultConditionsField.js]: Question: any reason we are using a p tag instead of readonly text inputs?

Comment (ScopeyNZ) [reply client/src/components/CKANResultConditionsField.js]: I considered that but it looked weirder than just not having it 🤷‍♂️ . I went over the designs with Sacha but _not_ what happens when you're looking at a read only version with nothing set.

**Comment by ScopeyNZ** [comment]
> I don't think the display conditions field is working. It's either not showing up at all for me or showing the label with an empty field content.

Hmmmn this is working for me. I can't reproduce any console errors.

File: client/src/components/CKANResultConditionsField.js (hunk 9 above)
Comment (ScopeyNZ) [reply client/src/components/CKANResultConditionsField.js]: Ok I've altered it now. The field on the PHP side will delegate to the existing handler when the value is empty.

File: src/Forms/ResourceLocatorField.php (hunk 8 above)
Comment (ScopeyNZ) [reply src/Forms/ResourceLocatorField.php]: Done

File: client/src/components/CKANResourceLocatorField.js [hunk 10]
```diff
@@ -280,18 +280,23 @@ class CKANResourceLocatorField extends Component {
...
-      return <Field {...sharedProps} type="text" disabled />;
+    // If we're not valid or read-only then we'll render just a simple disabled input in place of
+    // the select
+    if (readOnly || !currentDataset || !uri || !uri.length) {
+      let value = '';
+      if (spec && spec.resource && currentDataset) {
+        const resource = currentDataset.resources.find(candidate => candidate.id === spec.resource);
+        value = resource.name || resource.description || resource.id;
+      }
+      return <TextFieldComponent {...sharedProps} type="text" disabled value={value} />;
```
Comment (robbieaverill) [inline client/src/components/CKANResourceLocatorField.js]: Do you still need `type="text"` since it's a TextField now?

**Pull request merged**